rbc/external.py

```diff
@@ -2,106 +2,125 @@
 """
 
 # This code has heavily inspired in the numba.extending.intrisic code
-from typing import Dict, List, Union
-from collections import defaultdict
+import warnings
+from typing import List, Union
 from numba.core import extending, funcdesc, types, typing
 from rbc.typesystem import Type
 from rbc.targetinfo import TargetInfo
+from rbc.errors import UnsupportedError
+from rbc.utils import validate_devices
 
 
 class External:
+
     @classmethod
-    def external(cls, *args):
-        """
-        Parameters
+    def external(cls, signature, *, devices=['CPU']):
+        """Parameters
         ----------
         signature : object (str, ctypes function, python callable, numba function)
             Any object convertible to a Numba function via Type.fromobject(...).tonumba()
+
+        Keyword parameters
+        ------------------
+        devices : list
+          Device names ('CPU' or/and 'GPU') for the given set of
+          signatures. Default is ['CPU'].
         """
-        ts = defaultdict(list)
-        key = None
-        for signature in args:
-            with TargetInfo.dummy():
-                t = Type.fromobject(signature)
-            if not t.is_function:
-                raise ValueError("signature must represent a function type")
-
-            if not t.name:
-                raise ValueError(
-                    f"external function name not specified for signature {signature}"
-                )
-
-            if key is None:
-                key = t.name
-            if not key:
-                raise ValueError(
-                    f"external function name not specified for signature {signature}"
-                )
-
-            for device in [
-                a.upper() for a in t.annotation() or [] if a.upper() in ["CPU", "GPU"]
-            ] or [
-                "CPU",
-                "GPU",
-            ]:
-                ts[device].append(signature)
-
-        obj = cls(key, ts)
+        with TargetInfo.dummy():
+            t = Type.fromobject(signature)
+        if not t.is_function:
+            raise ValueError("signature must represent a function type")
+        name = t.name
+        if not name:
+            raise ValueError(
+                f"external function name not specified for signature {signature}"
+            )
+
+        # Default devices contains `CPU` only because in most cases
+        # the externals are defined for the CPU target.
+        devices = validate_devices(devices)
+        # Using annotation devices (specified via `| CPU` or `| GPU`) is deprecated.
+        # TODO: turn the warning to an exception for rbc version 0.12 or higher
```

Comment:
Could you create an issue for that?

Reply:
Done: https://github.com/xnd-project/rbc/issues/544